### ISSUE TITLE
Permanent install error "Lock file does not match YAML configuration"

### ISSUE THREAD

**Issue by Barberrrry** [Masterminds/glide#185]
I'm using glide 0.8.2 installed by homebrew on Mac OS.
I have the problem: _glide install_ always returns __Oops! Lock file does not match YAML configuration. Consider running 'update'__ even immediately after _glide up_.

**Comment by mattfarina** [comment]
When `glide up` creates a `glide.lock` file it generates a hash of the `glide.yaml` configuration and puts it in the `glide.lock` file. Running `glide install` generates a new has of the `glide.yaml` configuration and then checks it agains the value in the `glide.lock` file.

Is the content of your `glide.yaml` file changing anywhere?

If you run the following so you see the error:
```sh
$ glide up && glide --debug install
```

Can you share your project or `glide.yaml` file? I'd like to try and reproduce the problem.

**Comment by Barberrrry** [comment]
I don't make changes in glide.yaml between `glide up` and `glide install`. `glide up` successfully makes `glide.lock` and puts hash in it, as you said. If I make one more `glide up`, only `updated` field is changed, `hash` is still the same. Anyway, `glide install` continues to return error.

Unfortunately, I can't share glide.yaml with you, because there is a lot of private repos in it.

I will try to build glide binary file from source instead of using homebrew build and report about results.

**Comment by Barberrrry** [comment]
The same result is with Glide 0.8.2 builded from the source.

**Comment by mattfarina** [comment]
I created the branch `issue/185` with some addition debug code. If you run `glide up` and `glide install` now it will print out more information. Does something in there help you see the problem?

**Comment by Barberrrry** [comment]
I've tried out and hashes are different:
```[WARN] Lockfile hash: 2d75bf86c2d24e3bbf1da70e1a286d23bfa06ee4f8bb17b9d37f75548a83ac18
[WARN] glide.yaml generated hash: 36a281dd849ec9142b380a3199a5d9f04a6d5cad45c918d1d13c976acc661133```
I repeated `glide up` and `glide install` many times and these hash values were the same every time.

As I understand, hash of `glide.yaml` is calculated not from file content, but from struct Config decoded from YAML. Maybe Config is changing while `glide up` is working and these changes affect hash that is put to `glide.lock` finally. `glide install` open the file, decodes it to Config and reveals that the hash is different.

**Comment by mattfarina** [comment]
I just pushed another update with another line of debug code. This is important because it's the hash at write time which is a different point than the hash is generated for the lock.

I added `fmt.Println` calls to print out the `glide.yaml` representation of the `Config` instances when they are written on update and when the hashes are being compared on install. Do you see a difference in the yaml?

**Comment by Barberrrry** [comment]
Cool! I think, the problem is revealed.
The difference between two states of yaml is in version ranges of some packages. I have this in `glide.yaml`:
```
- package: bitbucket.org/lazadaweb/go-trace
  version: ~1.0
- package: bitbucket.org/lazadaweb/go-logger
  version: ~1.0
```
but Glide transform it to this while resolving version with other packages which use these 2 packages:
```
- package: bitbucket.org/lazadaweb/go-trace
  version: ~1.0, ~1.x
- package: bitbucket.org/lazadaweb/go-logger
  version: ~1.0, ~1.x, ~1.x, ~1.x
```
So `glide install` takes hash from first original version, whilst `glide up` calculates hash from the second version and puts it to `glide.lock`.

And of course, if I do `glide get <some_package>` file `glide.yaml` is automatically rewrited and I get the second version again.

**Comment by mattfarina** [comment]
Thanks for the detail! I know what's going on now and can fix this.

**Comment by mattfarina** [comment]
@Barberrrry I created the branch `issue/185-fix` (in pull request #187). Can you test this and see if it fixes your problem?

**Comment by Barberrrry** [comment]
Thank you for fix! I will check it as soon as possible and let you know
about results.

среда, 23 декабря 2015 г. пользователь Matt Farina написал:

> @Barberrrry <https://github.com/Barberrrry> I created the branch
> issue/185-fix (in pull request #187
> <https://github.com/Masterminds/glide/pull/187>). Can you test this and
> see if it fixes your problem?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Masterminds/glide/issues/185#issuecomment-166946922>.
>


-- 
С уважением,
Петров Вадим
vadim.petrov@gmail.com

**Comment by Barberrrry** [comment]
With `issue/185-fix` everything is alright, thank you! Which version would include this fix?

**Comment by mattfarina** [comment]
I have merged the fix into master and will be releasing 0.8.3 with it soon.
